### DISCUSSION TITLE
MQTT how do you make it go?

### DISCUSSION THREAD

**Discussion by OldSurferDude** [Dan-in-CA/SIP#293] [Q&A]
I have configured my 25 zone irrigation hardware to be controlled by MySensors / MQTT / Home Assistant.  At this point, all I can do is run a zone (station) by putting data into the _Zone_ topic, then the _Duration_ Topic, then _Start_, topic.  Now I need a UI for programing the irrigation and it looks like SIP will do it.

I just found SIP this week, so I'm almost at zero on the learning curve.  I see that I will need most of the MQTT plugins, but I can't figure out how to integrate them and configure them.  

- I see that MQTT requires Paho MQTT.  is that a matter of just running "pip install paho-mqtt"?  ("_Successfully installed paho-mqtt-1.6.1_"
- Next I'll need to integrate the MQTT plug in.  How is that done? ([PLUGINS]-->[BROWSE MORE PLUGINS]-->(check all MQTT plugins but slave)-->[INSTALL]-->(check Enabled on all MQTT plugins)-->[UPDATE ENABLED]

(note, I keep getting "SIP didn't restart.  Restart from command line"  is that a problem?)

Now how do I configure:
- MQTT broker?
- Which stations to send?
- What topics?
- integrated with Home Assistant?

I look forward to being a member of this community

-OSD

**Comment by Dan-in-CA** [comment]
Welcome!

For integration with home assistant the "mqtt_hass" plugin is for just that purpose. It requires running
"python3 -m pip install python-slugify --upgrade" before installing.

Do you have a MQTT broker installed on your network? If not Mosquitto is a good one:
https://mosquitto.org/download/

Once you have a broker available, You just need to add the URL of the broker to the "MQTT Broker Host:" field on the MQTT plugin setup page. Most of the other settings can use the defaults.

The "SIP didn't restart" message is not a problem It just means that SIP is running from the command line. When SIP tries to restart it relies on the "systemctl restart sip" command which is available when the systemd sip.service is installed.

Dan

**Reply by OldSurferDude** [reply]
Thanks, Dan!

By thrashing, I found that after the plugin is installed and enabled, one clicks on [PLUGINS] to configure.  For the _MQTT Broker Hosts_  I put **192.168.1.142**, which is the IP address of the computer that is hosting my MQTT broker.  Nothing has appeared on my broker.  But then, you wrote URL.  does that mean I should put **http://192.168.1.142:1883** in that field?

I appreciate your prompt response, and will definitely appreciate you patience as I'm going to be asking a lot of noob questions.  I have read the wiki and I was unable to connect some dots, thus my noob questions.

Thanks for your hellp!

OSD

**Comment by Dan-in-CA** [comment]
It should just be 192.168.1.142
the 1883 port # is in the MQTT Broker Port field.

**Reply by OldSurferDude** [reply]
Thanks for the confirmation, Dan!

I have a simple  Program to run the first Station for 1 minute at midnight.  I clicked on [RUN NOW].  No data appeared on the MQTT broker.  How does that data get there?

Once on the MQTT broker, how do I get Home Assistant to acknowledge that data?

**Comment by Dan-in-CA** [comment]
The mqtt_zones plugin broadcast information when a station is turned on or off. It sends a list of all stations ("zone_list":) represented by 0 for off stations and 1 for running stations and also a dictionary ("zone_dict") showing each station name and it's status 1 or 0.

Do you have a way to observe the MQTT traffic such as MQTT Explorer?

As far as the HA integration, I am not a HA user so I can't help with that.

**Comment by OldSurferDude** [comment]
When I click on [RUN NOW] I see the output from SIP (running from the command line on my RPi) is 
`192.168.1.227:59279 - - [09/Dec/2022 11:14:30] "HTTP/1.1 GET /rp" - 303 See Other
`

**Comment by Dan-in-CA** [comment]
That shows the traffic from the RPi to the system you are using to view SIP's web based user interface.
It does not show any of the MQTT traffic. You need a separate app for that like MQTT Esplorer:
http://mqtt-explorer.com/

**Reply by OldSurferDude** [reply]
Yes, I have MQTT explorer

Here's what I've done.  I disabled and uninstalled all the MQTT plugins.  I installed, enabled, MQTT base, then restarted SIP.  I configured MQTT base, then restarted SIP.  I did the same for MQTT: set, get, and zones.  Nothing appeared at the MQTT broker when I did a [RUN NOW].  I notice that in the MQTT base program user and MQTT Broker Username: and MQTT Broker Password was populated with "user" and "password",  I cleared the dialog boxes and saved.  Still no joy. :(

Any ideas?

**Comment by Dan-in-CA** [comment]
One thing that is not real obvious is on the MQTT_zones plugin setup page you need to click the "Submit" button at the bottom of the page before it will work.

**Comment by OldSurferDude** [comment]
Well, that didn't work, Dan :(

I had seen [SUBMIT], but just to be sure, I uninstalled the MQTT plugins and did the install, enable, configure (with [SUBMIT], but still no joy.  hmmmmmmm?  I know that the ip address and port are working because MQTT Explorer and MySensors are seeing it.  The ethernet connection is good because MySensors is running on the same RPi.  I checked that Paho-mqtt is installed.  I set the program to run automatically and that didn't result in data going to the MQTT broker.

I appreciate your help.

OSD

**Comment by Dan-in-CA** [comment]
In the left panel of MQTT Explorer do you see SIP or sip listed?
If so, what happens if you click on the arrows next to them"

Also, what version OS are you running on the Pi?

**Reply by OldSurferDude** [reply]
Hi Dan,

No, SIP did not appear in MQTT Explorer.  I have an MQTT broker specifically for this testing so that I can avoid any confusion.  Also running on this RPi is the MySensors gateway.

My OS on Raspberry Pi 3B+ is:
PRETTY_NAME="Raspbian GNU/Linux 11 (bullseye)"
NAME="Raspbian GNU/Linux"
VERSION_ID="11"
VERSION="11 (bullseye)"
VERSION_CODENAME=bullseye
ID=raspbian
ID_LIKE=debian
HOME_URL="http://www.raspbian.org/"
SUPPORT_URL="http://www.raspbian.org/RaspbianForums"
BUG_REPORT_URL="http://www.raspbian.org/RaspbianBugs"

I am running the RPi headless, that is, Raspberry Pi OS Lite (32-bit)

The lack of communication to the broker occurs whether the MySensors gateway is running or not.

I am running SIP from the command line

While you're scratching your knowledgeable head, I'll create an desktop RPi Desktop

Thanks for all the scratching!

OSD

**Comment by OldSurferDude** [comment]
I flashed a new image of Raspberry Pi OS Lite (32-bit) Bullseye on an SD card from SanDisk.

git was not preinstalled. 
```
sudo apt update --fix-missing
sudo apt install git
```
then the install per the wiki.   Then I checked to see if this worked and set the software's webserver IP address . Then I terminated SIP to begin install the components for MQTT.  Each step of the way the checked to see if SIP still worked.
```
sudo apt install python3-pip
pip install paho-mqtt
pip install python-slugify --upgrade
```
Then I installed and enabled the MQTT plugin, restarted SIP, configured the MQTT plugin, restarted SIP

At this point, MQTT Explorer shows a SIP topic with data of "UP" (or "DOWN" if SIP is not running)

Installed, enabled, configured, the other MQTT plugins and I have checked, with a simple Program, that Zones is working as expected.

Now I must express my embarrassment by pointing out that my problem was probably an **inadequate or bad SD card**.  I have learned a hard lesson, unfortunately at your expense.  I appreciate the time you took and your advice was spot on.  Alas, my problem was hardware, not software.

Sincerely, thank you!

OSD

**Comment by Dan-in-CA** [comment]
Fantastic!
Glad it's working now.
As far as integrating SIP with Home Assistant; 
If you install the MQTT_hass plugin and open the set up page from the Plugin menu then click the Help button you will be taken to a page with detailed information about how to set it up.

Yes I was scratching my head about your problem but I got some new dandruff shampoo and that should help.

Dan

**Reply by OldSurferDude** [reply]
ARRRRRRRGGGGGG!

I need to document exactly how this is all put together.  So I start over from scratch.  At the first login, first prompt, of a virgin install of RPi lite, I executed the curl-automatic install.  I had to install python3-pip and then pip install paho-mqtt.  no joy.

With the desktop version, in a terminal, the curl-automatic install, python3-pip was already installed, pip install paho-mqtt. it seems to work.

I'm doing more testing.

Any thoughts, Dan?

Note: the curl-automatic does not "sudo systemctl enable sip"  This needs to be done before rebooting.

**Reply by OldSurferDude** [reply]
Things are now working pretty well.  Here's what I discovered

`The command line version does not have all the necessary libraries.  These commands must be run before anything else
```
sudo apt update --fix-missing -y
sudo apt install python3-pip -y
```
The paho-mqtt module has to be installed with _sudo_ and before the install script
`**sudo** pip install paho-mqtt

By doing this the sip.service appears to be starting at boot, so that's good, too :)

I'm still testing it, but it looks like the generic, off-brand SD card is a no-go.

So un-arrggg

Hope this helps someone else.  

Onward

OSD
